File: Top_Interview_Questions/courseScheduleII.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adjList = [[] for _ in range(numCourses)]
        in_degree = [0] * numCourses
        for c1, c2 in prerequisites:
            # c2-> is edge
            adjList[c2].append(c1)
            in_degree[c1] += 1
        
        q = collections.deque([])

        for ci in range(numCourses):
            if in_degree[ci] == 0:
                q.append(ci)

        topo_ordering = []
        while q:
            ci = q.popleft()
            topo_ordering.append(ci)
            for neighbor in adjList[ci]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    q.append(neighbor)

        if len(topo_ordering) != numCourses:
            return []
        return topo_ordering

File: NeetCode_Roadmap/course_schedule_II.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        int N = numCourses;
        vector<vector<int>> adj_list(N, vector<int>());
        vector<int> indegree(N, 0);
        for (const auto& vec : prerequisites) {
            int a = vec[0], b = vec[1];
            adj_list[b].push_back(a);
            ++indegree[a];
        }
        std::queue<int> q;
        for (int n = 0; n < N; ++n) {
            if (indegree[n] == 0) q.push(n);
        }
        std::vector<int> topo_sort;
        while (!q.empty()) {
            int top = q.front();
            q.pop();
            topo_sort.push_back(top);
            for (const auto neighbor : adj_list[top]) {
                if (--indegree[neighbor] == 0) q.push(neighbor);
            }
        }
        if (topo_sort.size() != N) return vector<int>();
        return topo_sort;
    }
};